content/en/blog/_posts/2024-04-15-devsecops-in-action-k8s-secrets.md

```diff
@@ -0,0 +1,562 @@
+			isExist = true
+		}
+	}
+}
+```
+
+Now applications check `DEMO_SECRET__PASSWD_FILE` was defined then application
+checks whether the file exists by value from the environment variable.
+
+*[kustomization/overlays/demo/kustomization.yml](https://github.com/efrikin/devsecops-in-action-kubernetes-secrets/blob/96afd59f8a391973a8ed5a9b410915b9e99d4bf4/kustomization/overlays/demo/kustomization.yml#L25-L45)*
+```yaml
+...
+spec:
+  containers:
+  - envFrom:
+    - secretRef:
+         name: demo-secret
+    volumeMounts:
+    - mountPath: /secrets
+      name: demo-secret
+      readOnly: true
+  volumes:
+  - name: demo-secret
+    secret:
+      optional: true
+      secretName: demo-secret
+...
+```
+
+`spec.containers.0.envFrom` defines all of the Secret's data as container
+environment variables and `spec.volume/spec.containers.0.volumeMounts`
+defines all mountpoints to Pod.
+
+Excellent! Part of recommendations have been completed, but for the full picture
+it's necessary to solve the following issues:
+
+* The secret file is a mountpoint and it's necessary to remove after it usage
+(e.g reading)
+* Kind Secret is insecure that's why it's necessary to minimize its use
+* Additional implementation of environment variable such as:
+`DEMO_SECRET__PASSWD_FILE` can Impede applications update.
+
+## Stage-02: Injection of secret to Pod instead of mountpoint
+
+In order to get rid of Secret as mountpoint it's necessary to use Injection of
+Secret via file to Pod. Often secret manager tools perform these functions.
+Also they provide functions such as: dynamic secrets, database credential
+rotation, automated PKI infrastructure, data encryption and tokenization,
+identity-based access, key management, etc.
+
+In the example below, simulation of getting secret from secure storage and
+delivery to Pod will be done via `spec.initContainers` then
+`spec.containers.0.postStart` hook will clean the file after successful passing
+of readiness probe in order to make file with sensitive data to Pod useless for
+attacker.
+
+*[kustomization/overlays/demo/kustomization.yml](https://github.com/efrikin/devsecops-in-action-kubernetes-secrets/blob/fba0bf048d1cd915fb265fc1b89041582a1d76f9/kustomization/overlays/demo/kustomization.yml#L23-L61)*
+```yaml
+...
+  initContainers:
+    - name: secret-injector
+      image: alpine:3.19
+      command:
+        - sh
+        - -c
+        - cat /dev/urandom | base64 -w 0 | fold -w 16 | head -1 > ${DEMO_SECRET__PASSWD_FILE}
+      envFrom:
+         -  secretRef:
+             name: demo-secret
+      volumeMounts:
+        - mountPath: /secrets
+          name: demo-secret
+...
+    lifecycle:
+      postStart:
+        exec:
+          command:
+          - sh
+          - -c
+          - wget -qO- localhost:8080/readiness && echo > ${DEMO_SECRET__PASSWD_FILE}
+```
+
+The container must change status to `READY` after the application reads a file
+with secret, otherwise `spec.containers.0.postStart` hook will clean it.
```

Comment:
Will clean it or won't clean it? If I understand correctly, Pod reads the Secret file, starts up correctly, then the postStart hook kicks in. Is that right?

Reply:
> Will clean it or won't clean it? If I understand correctly, Pod reads the Secret file, starts up correctly, then the postStart hook kicks in. Is that right?

Right. If file was injected to Pod.